src/com/google/enterprise/adaptor/ValidatedUri.java

```diff
@@ -0,0 +1,115 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.enterprise.adaptor;
+
+import com.google.common.base.Strings;
+import java.io.IOException;
+import java.net.InetAddress;
+import java.net.MalformedURLException;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.net.URL;
+import java.util.concurrent.TimeUnit;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+/**
+ * Validates URIs by syntax checking and validating the host is reachable.
+ */
+public class ValidatedUri {
+  /** The logger for this class. */
+  private static final Logger log =
+      Logger.getLogger(ValidatedUri.class.getName());
+
+  /** The connect timeout, in milliseconds. */
+  private static final int TIMEOUT_MILLIS = (int) TimeUnit.SECONDS.toMillis(30);
+
+  /** The validated URI. */
+  private final URI uri;
+
+  /**
+   * Attempts to validate the given URI syntax with some more stringent checks
+   * than new URI. In this case, we're mostly trying to catch typos and obvious
+   * configuration issues.
+   *
+   * @param uriString the URI to test
+   * @throws URISyntaxException if the URL syntax is invalid
+   */
+  public ValidatedUri(String uriString) throws URISyntaxException {
+    if (Strings.isNullOrEmpty(uriString)) {
+      throw new URISyntaxException("" + uriString, "null or empty URI");
+    }
+    try {
+      // Basic syntax checking, with more understandable error messages.
+      // Also ensures the URI is a URL, not a URN.
+      new URL(uriString);
+      // Advanced syntax checking, with more cryptic error messages.
+      uri = new URI(uriString);
+      uri.toURL();
+    } catch (MalformedURLException e) {
+      int index = e.getMessage().indexOf(": ");
+      String reason;
+      if (index > 0 && index == e.getMessage().lastIndexOf(": " + uriString)) {
+        reason = e.getMessage().substring(0, index);
+      } else {
+        reason = e.getMessage();
+      }
+      throw new URISyntaxException(uriString, reason);
+    }
+
+    if (!uri.isAbsolute()) {
+      throw new URISyntaxException(uriString, "relative URIs are not allowed");
+    }
+
+    if (Strings.isNullOrEmpty(uri.getHost())) {
+      throw new URISyntaxException(uriString, "no host");
+    }
+
+    if (Strings.isNullOrEmpty(uri.getRawPath())
+        && Strings.isNullOrEmpty(uri.getRawQuery())
+        && Strings.isNullOrEmpty(uri.getRawFragment())) {
+      throw new URISyntaxException(uriString,
+          "no path, query, or fragment components");
```

Comment:
s/or/and/

or maybe "path, query, and fragment are all missing".  Would that be true for "http://www.google.com"?   What makes this URL invalid?